feat(api): poll module error states and raise them

sfoster1 opened the pull request:
Use the new module gcodes that we recently added for heater-shaker and thermocycler to poll for module error status and toss it in the error callback.

## Other Cool Refactors
- The gcode tests need to know about the new polled gcode and not diff it 76f236197948abe8236791c510eec043d2b4b84a
- The modules need to not just turn themselves off after one error a5fa1a676638ae7201f6acfbfd311c9ac9b1adda
- The modules say they're in an error state over and over, so
   - We need to swallow repeated errors in the engine; if we don't suppress errors in the engine once we're in an error state, then we'll `finish()` over and over and make a nice disco from the z axis brakes 1beb9f9f35523fe65cbfb10fb8fe37e1043ecb47
   - We need to log gcode based errors in pollers less verbosely or we blow out the log c2c7e7058ea8f5cb3aa17f182ae213a383005d5b
- These commands have multiple acks! This will make the response parsers get out of sync forever. Fix this by looking for, well, two acks 6bf9a46f5b31f712abc76fd7e069e2d0e5eab8a8
- We need to not just step on the run error when we autoclear and pretend it's an estop a11f4203a089e5b724394b18fbcb3b24c525c0e4
- The module emulators were totally hosed and now they're less hosed 3bf832c135d5dbf5d54a20ded1edc2be08be54e2

All in all a good tour of the lesser-touched parts of our stack.

Closes EXEC-1703

## tests
- [x] connect a heater/shaker; push a version that sets a delayed error when you set a temperature; check that the async error happens
- [x] ditto thermocycler

sfoster1 committed:
feat(api): poll module error states and raise them

Use the new module gcodes that we recently added for heater-shaker and
thermocycler to poll for module error status and toss it in the error
callback.

Closes EXEC-1703

ryanthecoder reviewed (approved):
lgtm

vegano1 reviewed (approved):
Looks good, thank you.

sfoster1 committed:
fix: update the gcode tests

A couple problems here
- They need to know about the new gcodes
- They need to know these are polling gocdes and should be ignored
- The supporting code for updating the http tests is now (was always?)
broken and needed to be awaited

sfoster1 committed:
handle module error callbacks in reader

sfoster1 committed:
ok definitely do not do that

sfoster1 committed:
chore: shorter logging for polled gcode errors

sfoster1 committed:
refactor: handle multi-ack commands in serial

The new error status commands in the module have multiple acks in
response to the command itself. In the current implementation, the
serial machinery reads only one command response, which means that the
command after an M411 would get the second ack to the M411, and the
command after that would get the response to the command that got the
M411 response... it's off by one forever.

To fix this we can add a a new serial communication method that handles
these kinds of commands, backed by some extra logic that is also a nice
clean refactor to the serial machinery.

sfoster1 committed:
refactor(api): swallow repeated module exceptions

Module errors are states, which means we'll keep telling the engine
about them as long as the error state persists. To avoid calling
finish() over and over, swallow module errors if the engine is already terminal.

sfoster1 committed:
refactor: do not override run errors with estop

This was around to handle the case where the engine was stopped by an
asynchronous estop call; we wouldn't get the estop error and the run
would look successful, so if the engine was stopped asynchronously then
we'd override with an estop error. But now there's multiple asynchronous
error sources, so only do the override if we were going to finish cleanly.

sfoster1 committed:
refactor: fix module emulator tests

Couple fun problems here such as
- All the stm32 modules were just using the wrong ack all the time and
the wrong response patterns, which mostly happened to work
- They didn't echo the command back either
- The thermocycler emulator was awkwardly halfway in between a gen1 and
gen2 in terms of what's implemented, and couldn't switch between the two
based on the hardware version it was told
- Also the hardware model that you put into the emulator config isn't in
any way related to what the equivalent hardware will tell you so it's a
good thing we don't look at the result ever

Fix these things by
- Implementing the new module response style, parametrizing some stuff
in the emulator responses along the way
- The thermocycler will try and guess what version you meant; please put
an actual model in there like what the software expects
- It will then actually behave like that version, which is new and bold
- That means our emulator tests now get thermocycler gen2 devices which
is cool since ~0 gen1s exist in the wild

sfoster1 committed:
one (hopefully) last fixup

The thermocycler emulator for some reason doesn't send responses
sometimes but it should always ack. Also we always want the
thermocyclers to be gen1 in the gcode tests because that's what they
were written for.

vegano1 commented on api/src/opentrons/hardware_control/emulation/heater_shaker.py:
```diff
@@ -145,6 +146,14 @@ def _deactivate_heater(self, command: Command) -> str:
-    @staticmethod
-    def get_terminator() -> bytes:
+    def _get_error_state(self, command: Command) -> str:
+        return f"M411 {HS_ACK}M411"
```
Should this be `f"M411 {HS_ACK}"` ?

sfoster1 replied:
No, this is for sending the double ack (one from the motor, one from the heater) that this sends from the hardware. the second `HS_ACK` is provided by the code that makes sure every command response has the ack. So think of it less as `M411 {HS_ACK}` with an `M411` on the end, and more as an `M411` with an additional `M411 {HS_ACK}` added beforehand
